### PR TITLE
Replace --prefer-shared-library with --build-shared-library

### PR DESCRIPTION
**Pull request by cbracken**
This replaces the --prefer-shared-library flag, which falls back to
regular (non-shared-lib) compile if the NDK is not found, with the
--build-shared-library flag, which exits with an error message if the
NDK is not found.

This simplifies the set of allowed code paths through AOT compile,
resulting in better testability and easier-to-follow logic. It also
results in more predictable behaviour for continuous integration and
other scenarios.

### CODE REVIEW

**Comment by cbracken** [comment]
Related #17236. Specifically [this comment](https://github.com/flutter/flutter/issues/17236#issuecomment-386410915).

**Comment by mraleph** [comment]
@cbracken I don't think anybody uses this flag. We added this flag to enable profiling with native tools (simpleperf) instead of Observatory on Android. Given this we don't need any deprecation period AFAIK.

File: packages/flutter_tools/lib/src/base/build.dart [hunk 1]
```diff
@@ -148,16 +148,19 @@ class AOTSnapshotter {
...
+      printError('Could not find NDK compiler. Unable to build with --build-shared-library');
+      return -2;
     }
 
     final PackageMap packageMap = new PackageMap(packagesPath);
     final String packageMapError = packageMap.checkValid();
     if (packageMapError != null) {
       printError(packageMapError);
-      return -2;
+      return -3;
```
Comment (mraleph) [inline packages/flutter_tools/lib/src/base/build.dart]: What are these magic numbers? Should they have a named constant declared somewhere for them?

File: packages/flutter_tools/lib/src/base/build.dart [hunk 2]
```diff
@@ -148,16 +148,19 @@ class AOTSnapshotter {
     // TODO(cbracken): replace IOSArch with TargetPlatform.ios_{armv7,arm64}.
     assert(platform != TargetPlatform.ios || iosArch != null);
 
-    final bool compileToSharedLibrary = preferSharedLibrary && androidSdk.ndkCompiler != null;
-    if (preferSharedLibrary && !compileToSharedLibrary) {
-      printStatus('Could not find NDK compiler. Not building in shared library mode.');
+    // buildSharedLibrary is Android-only.
+    assert(platform != TargetPlatform.ios || !buildSharedLibrary);
```
Comment (mraleph) [inline packages/flutter_tools/lib/src/base/build.dart]: what happens if you run with asserts disabled, build for iOS and we pass buildSharedLibrary by accident? 

can we simply make it a noop on iOS instead of an assertion?

File: packages/flutter_tools/lib/src/commands/build_aot.dart [hunk 3]
```diff
@@ -33,11 +33,16 @@ class BuildAotCommand extends BuildSubCommand {
         hide: !verboseHelp,
         help: 'Preview Dart 2.0 functionality.',
       )
+      ..addFlag('build-shared-library',
+        negatable: false,
+        defaultsTo: false,
+        help: 'Compile to a *.so file (Android only; requires NDK).'
```
Comment (mraleph) [inline packages/flutter_tools/lib/src/commands/build_aot.dart]: maybe make it work on both iOS and Android - it will be a NOP on iOS. 

You can then change this comment to `" (Requires NDK when building for Android)"`

File: packages/flutter_tools/lib/src/base/build.dart [hunk 4]
```diff
@@ -148,16 +148,19 @@ class AOTSnapshotter {
...
+
+    if (buildSharedLibrary && androidSdk.ndkCompiler == null) {
```
Comment (mraleph) [inline packages/flutter_tools/lib/src/base/build.dart]: do you plan to make `ndkCompiler` easy to configure with an additional option or this would be a part of a separate PR?

I think current error is not actionable - which makes it confusing. Consider morphing it into:

`Could not find NDK compiler at ${ndkPath}. Unable to build with --build-shared-library. Configure ${ndkPath} by setting --ndk-path flag`

or something a long these lines

**Review by mraleph — APPROVED** [review]
LGTM with comments addressed

File: packages/flutter_tools/lib/src/base/build.dart (hunk 1 above)
Comment (cbracken) [reply packages/flutter_tools/lib/src/base/build.dart]: We should probably just replace them with a proper failure mode. They were initially there for our own use to differentiate the different types of failure. Those should be evident with proper trace logging either way.

Comment (cbracken) [reply packages/flutter_tools/lib/src/base/build.dart]: For now just replaced these with a uniform failed exit code. We should probably just move to throwToolExit like everywhere else since we never really use the exit codes other than to log error messages. Will do that in a followup.

File: packages/flutter_tools/lib/src/base/build.dart (hunk 2 above)
Comment (cbracken) [reply packages/flutter_tools/lib/src/base/build.dart]: Done.

File: packages/flutter_tools/lib/src/commands/build_aot.dart (hunk 3 above)
Comment (cbracken) [reply packages/flutter_tools/lib/src/commands/build_aot.dart]: Done.

File: packages/flutter_tools/lib/src/base/build.dart (hunk 4 above)
Comment (cbracken) [reply packages/flutter_tools/lib/src/base/build.dart]: I've updated the error message to be more actionable. For the moment we continue to look for the NDK bundle in the SDK, but we can followup with a flag that allows an override if that's something that would be useful.
